sample-code/src/server.ts

```diff
@@ -265,3 +273,95 @@ app.get('/langchain/invoke-rag-chain', async (req, res) => {
       .send('Yikes, vibes are off apparently 😬 -> ' + error.message);
   }
 });
+
+/* Document Grounding */
+app.get(
+  '/document-grounding/invoke-orchestration-grounding',
+  async (req, res) => {
+    try {
+      res.setHeader('Content-Type', 'text/event-stream');
+      res.setHeader('Connection', 'keep-alive');
+      res.flushHeaders();
+
+      // Create an empty collection.
+      const collectionId = await createCollection();
+      res.write(`Collection created:\t\t\t${collectionId}\n`);
+
+      // Create a document with the current timestamp.
+      const timestamp = Date.now();
+      await createDocumentsWithTimestamp(collectionId, timestamp);
+      res.write(`Document created with timestamp:\t${timestamp}\n`);
+
+      // Send an orchestration chat completion request with grounding module configured.
+      const groundingResult = await orchestrationGrounding();
+      res.write(
+        `Orchestration responded with timestamp:\t${groundingResult.getContent()}\n`
+      );
+
+      // Delete the created collection.
+      await deleteCollection(collectionId);
+      res.write(`Collection deleted:\t\t\t${collectionId}\n`);
+
+      res.end();
+    } catch (error: any) {
+      console.error(error);
+      res
+        .status(500)
+        .send('Yikes, vibes are off apparently 😬 -> ' + error.message);
+    }
+  }
+);
+
+app.get('/document-grounding/invoke-retrieve-documents', async (req, res) => {
+  try {
+    res.setHeader('Content-Type', 'text/event-stream');
+    res.setHeader('Connection', 'keep-alive');
+    res.flushHeaders();
+
+    // Create an empty collection.
+    const collectionId = await createCollection();
+    res.write(`Collection created:\t\t\t${collectionId}\n`);
+
+    // Create a document with the current timestamp.
+    const timestamp = Date.now();
+    await createDocumentsWithTimestamp(collectionId, timestamp);
+    res.write(`Document created with timestamp:\t${timestamp}\n`);
+
+    // Retrieve documents directly from document grounding service.
+    const retrievalResult = await retrieveDocuments();
+
+    console.log(JSON.stringify(retrievalResult));
+
+    res.write('Retrieved documents:\n');
+    (retrievalResult.results as RetievalPerFilterSearchResult[]).forEach(
+      perFilterSearchResult => {
+        res.write(`  - Filter: ${perFilterSearchResult.filterId}\n`);
+        perFilterSearchResult.results!.forEach(
+          retievalDataRepositorySearchResult => {
+            res.write(
+              `    - Data repository: ${retievalDataRepositorySearchResult.dataRepository.title}\n`
+            );
+            retievalDataRepositorySearchResult.dataRepository.documents.forEach(
+              retrievalDocument => {
+                retrievalDocument.chunks.forEach(chunk => {
+                  res.write(`      - Chunk: ${chunk.content}\n`);
```

Comment:
This is a nice learning for me, I didn't think about accessing the results like this in a nested structure when I initially wrote an E2E test for the same case :) 
Looks cumbersome though, would be nice to have convenience functions to also access relevant results.

Reply:
Definitely agree. This API is cumbersome and we could provide some convenient functions. But I guess it is what it is due to the modeling of retrieval API.